[SPARK-22111][MLLIB] OnlineLDAOptimizer should filter out empty documents beforehand

akopich opened the pull request:
## What changes were proposed in this pull request?

The empty documents are filtered out in the `initialize` method and are never included in mini-batches.  `batchSize`, and `nonEmptyDocsN` are now the same thing.

## How was this patch tested?

Existing unit-tests.

akopich commented:
@WeichenXu123, @hhbyyh, looking forward to your opinion.

SparkQA commented:
**[Test build #83015 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83015/testReport)** for PR 19565 at commit [`721f235`](https://github.com/apache/spark/commit/721f235934f26e75172d39f0398365606616267f).

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -415,7 +415,8 @@ final class OnlineLDAOptimizer extends LDAOptimizer with Logging {
-    this.corpusSize = docs.count()
+    this.docs = docs.filter(_._2.numNonzeros > 0) // filter out empty documents
+    this.corpusSize = this.docs.count()
     this.vocabSize = docs.first()._2.size
```
Needs to be `this.docs` too

akopich replied:
`docs` is assumed to be non-empty, while `this.docs` may be empty. In this case `first()` fails.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -497,40 +495,38 @@ final class OnlineLDAOptimizer extends LDAOptimizer with Logging {
    * Update lambda based on the batch submitted. batchSize can be different for each iteration.
    */
-  private def updateLambda(stat: BDM[Double], batchSize: Int): Unit = {
+  private def updateLambda(stat: BDM[Double], batchSize: Double): Unit = {
```
Don't think this should change here

akopich replied:
I've done this just in order to achieve consistency with `updateAlpha`. Is it a bad idea?

akopich replied:
Guess, isn't. If we want this kind of consistency, `batchSize` should rather be `Double` all the way. So I've change the type to `Long`. Thanks for the comment.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -497,40 +495,38 @@ final class OnlineLDAOptimizer extends LDAOptimizer with Logging {
     // Update lambda based on documents.
     lambda := (1 - weight) * lambda +
-      weight * (stat * (corpusSize.toDouble / batchSize.toDouble) + eta)
+      weight * (stat * (corpusSize.toDouble / batchSize) + eta)
```
This isn't necessary to achieve floating-point division.

srowen reviewed (changes requested):
That looks pretty reasonable

akopich committed:
[SPARK-22111][MLLIB] OnlineLDAOptimizer should filter out empty documents beforehand

## What changes were proposed in this pull request?

The empty documents are filtered out in the `initialize` method and are never included in mini-batches.  `batchSize`, and `nonEmptyDocsN` are now the same thing.

## How was this patch tested?

Existing unit-tests.

SparkQA commented:
**[Test build #83015 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83015/testReport)** for PR 19565 at commit [`721f235`](https://github.com/apache/spark/commit/721f235934f26e75172d39f0398365606616267f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83015/
Test PASSed.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -497,40 +495,38 @@ final class OnlineLDAOptimizer extends LDAOptimizer with Logging {
+    val (statsSum: BDM[Double], logphatOption: Option[BDV[Double]], batchSize: Long) = stats
       .treeAggregate((BDM.zeros[Double](k, vocabSize), logphatPartOptionBase(), 0L))(
         elementWiseSum, elementWiseSum
       )
```
Not a big deal, but I think here use `mapPartition`(do aggregate in it) and then do treeAggregate, the process can be merged into one `treeAggregate`, is it right? it make the code more clear.

akopich replied:
Yes, that's right. But on the other hand, we've currently got a rather simple `treeAggregate` so it barely contributes to the complexity of the code. 

Anyway, is it OK to fix this in this issue?

WeichenXu123 replied:
hmm... I think it only need to move the code block in `mapPartition` into the aggregate `seqOp` function. Will it make code more complicated ?
I think it can be fix in this PR, it is only a simple modification and do not affect the logic.

akopich replied:
Actually, the block from `mapPartition` will be simplified, since it will no longer need to process collections of documents.

akopich replied:
It looks better now. Thank you for the suggestion.

akopich committed:
[SPARK-22111][MLLIB] OnlineLDAOptimizer should filter out empty documents beforehand

`mapPartition` and `treeAggregate` are merged together to a single `treeAggregate`.

SparkQA commented:
**[Test build #83040 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83040/testReport)** for PR 19565 at commit [`40685ee`](https://github.com/apache/spark/commit/40685ee960ef2c0e26b31ef96d1f3c8c974d3851).

SparkQA commented:
**[Test build #83040 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83040/testReport)** for PR 19565 at commit [`40685ee`](https://github.com/apache/spark/commit/40685ee960ef2c0e26b31ef96d1f3c8c974d3851).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83040/
Test PASSed.

akopich commented:
Now I feel that filtering empty docs out in the `initialize` is not a good idea, because it will be performed as many times, as the number of times `sample` in `next` gets called. Right?

Alternatively we can cache `this.docs`, but it's a waste of space...

akopich commented:
Or (and I think it would be the most efficient approach) we can just stick in the check for emptiness of the document to the `seqOp` of `treeAggregate`. However, it doesn't look like "filtering out beforehand". So, would this be OK?

SparkQA commented:
**[Test build #83045 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83045/testReport)** for PR 19565 at commit [`1d923f5`](https://github.com/apache/spark/commit/1d923f5908fd3a7a2ba8fa284d909aeb914ebc0a).

akopich committed:
[SPARK-22111][MLLIB] OnlineLDAOptimizer should filter out empty documents beforehand

batchSize should be Long.

SparkQA commented:
**[Test build #83045 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83045/testReport)** for PR 19565 at commit [`1d923f5`](https://github.com/apache/spark/commit/1d923f5908fd3a7a2ba8fa284d909aeb914ebc0a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83045/
Test PASSed.

akopich committed:
[SPARK-22111][MLLIB] OnlineLDAOptimizer should filter out empty documents beforehand

Useless check for an empty butch is removed. The emptiness of the batch is checked in `submitMiniBatch` anyway.

SparkQA commented:
**[Test build #83058 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83058/testReport)** for PR 19565 at commit [`8b7f30b`](https://github.com/apache/spark/commit/8b7f30bdb11abcd3efe2204c614695b286c0ae95).

SparkQA commented:
**[Test build #83058 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83058/testReport)** for PR 19565 at commit [`8b7f30b`](https://github.com/apache/spark/commit/8b7f30bdb11abcd3efe2204c614695b286c0ae95).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83058/
Test PASSed.

hhbyyh commented:
I wonder if we should add cache() for lda training data, even not for this feature. 

@srowen Not sure where we're on caching the training data or not for different algorithms. Appreciate your advice.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -446,14 +445,14 @@ final class OnlineLDAOptimizer extends LDAOptimizer with Logging {
   override private[clustering] def next(): OnlineLDAOptimizer = {
     val batch = docs.sample(withReplacement = sampleWithReplacement, miniBatchFraction,
       randomGenerator.nextLong())
-    if (batch.isEmpty()) return this
```
We still need this, right?

akopich replied:
I believe, it's redundant now. Anyway, `submitMiniBatch` counts the documents.

hhbyyh replied:
If that's the case, you should remove the comments below. 
"The methods assumes no empty documents are submitted."

akopich replied:
The method assumes no empty documents are submitted. But it's perfectly fine to submit an empty RDD. 

And I need to fix a typo.

hhbyyh replied:
I would prefer to leave the check here
1. performance-wise, sending an empty rdd to go through submitMiniBatch does not make much sense.
2. it will make submitMiniBatch more simple if no empty RDD will be sent, especially when we're filtering empty documents.

akopich replied:
So technically, you suggest to move the check 

```
if (batchSize == 0) {
      logWarning("No non-empty documents were submitted in the batch.")
      // Therefore, there is no need to update any of the model parameters
      return this
    }
```

to `next()`? Sounds reasonable.

akopich replied:
But performance wise... Currently we have a single pass. But if we move the check to `next`, `isEmpty` being an action will trigger the materialization of `batch` and then `batch` will be materialized once again when `treeAggregate` is called in `submitMiniBatch`. 

Right?

hhbyyh replied:
I think isEmpty() is optimized to avoid a full materialization.

hhbyyh replied:
If still you want to remove the line, would you please add a unit test to ensure `submitMiniBatch` can handle empty rdd? Thanks.

akopich replied:
I've added the test. Thank you.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -446,14 +445,14 @@ final class OnlineLDAOptimizer extends LDAOptimizer with Logging {
    * Submit a subset (like 1%, decide by the miniBatchFraction) of the corpus to the Online LDA
    * model, and it will update the topic distribution adaptively for the terms appearing in the
    * subset.
+   * The methods assumes no empty documents are submitted.
```
Maybe add a require

akopich replied:
Thank you. I'll add it.

hhbyyh reviewed (commented):
Some initial comments. Let's figure out the general approach first.

WeichenXu123 commented:
@akopich If you want to cache the input dataset, create JIAR to discuss it first. It's another issue I think. This JIAR also related to input caching issues: https://issues.apache.org/jira/browse/SPARK-19422

srowen commented:
Regarding caching: I think that can be ignored for purposes of this change. All this does is add a filter, and it doesn't cause an RDD to computed more than it was before.

The only question is whether the filtering is worth it; does it filter out enough that it makes later work faster?

akopich commented:
I am sure that caching may by avoided here. Hence, it should not be used. 

@srowen, maybe I don't get something, but I'm afraid, that currently lineage for a single mini-batch submission looks like this 
`docs.filter(nonEmpty).sample(minibatchFraction).treeAggregate(...)`. 

And I'm afraid that for each mini-batch `filter` will be performed all over again. But if we have smth like 
`docs.sample(minibatchFraction).filter(nonEmpty).treeAggregate(...)`,
this will be avoided. And no caching is needed.

WeichenXu123 commented:
@akopich IMO the filter won't cost too much, don't worry about the performance. (Or you can make a test to make sure)

srowen commented:
Filtering after sampling makes more sense. Though sampling isn't deterministic, it doesn't change the probability that any particular sample is produced.

akopich commented:
@WeichenXu123, yes there indeed is a difference in logic. Eventually it boils down to semantics of `miniBatchFraction`. If it is a fraction of non-empty documents being sampled, the version with `filter` going first is correct. If it's a fraction of documents (empty and non-empty) being sampled, then the version with `sample` going first is correct. To me the first version seems more reasonable (who cares about empty docs anyway). @srowen, if I get it right, you would prefer the second option. Why? 

@WeichenXu123, I agree with you: filtering introduces a minimal overhead. 

@srowen, regarding performance... I don't actually think it makes any difference unless complexity of `sample` depends on the length of the parent RDD. In all the subsequent computations empty documents can be handled effectively.

srowen commented:
I assume not-selecting a record in a sample is cheaper than just about any other operation, including filtering on a predicate. All else equal, I'd rather sample, then evaluate a predicate on only the sample.

The two versions produce the same result though. Either way, every x% sample of non-empty docs appears with equal probability.

Yes, I doubt one is meaningfully faster than the other overall though. If that's not the motive though, and there's not a functional change, what's the purpose here?

akopich commented:
Consider the following scenario. Let `docs` be an RDD containing 1000 empty documents and 1000 non-empty documents and let `miniBatchFraction = 0.05`.

Assume, we use `filter(...).sample(...)`. Then the resulted RDD will have around `50` elements. 

If we use `sample(...).filter(...)` instead, the `sample` returns around `100` elements. Now the number of elements in the RDD returned by `filter` is normally distributed. The expectation is `50` again though. 
Do I miss smth?

However, for larger samples this shouldn't make any difference. 

On the purpose of the issue: there were two different variables `batchSize`, and `nonEmptyDocsN` which could not be used interchangeably. The purpose is to submit a batch containing no empty docs which makes the two  variables refer the same value.

srowen commented:
I agree, they're the same. You said at https://github.com/apache/spark/pull/19565#issuecomment-339638791 that they weren't.

But if you're saying the code already filters out empty docs further upstream anyway, then there is no change in logic, just where the filtering happens. Or did I misunderstand that part?

akopich commented:
I'm saying they are not the same, but for larger datasets this should not matter.

There is a change in logic. The hack with 
`val batchSize = (miniBatchFraction * corpusSize).ceil.toInt` 
is not used anymore. The function `updateLambda` uses the real number of non-empty docs.

akopich commented:
And the empty docs were not explicitly filtered out. They've just been ignored in `submitMiniBatch`.

WeichenXu123 commented:
Yes, it changed the probability of samples indeed compared with current code.
But according to the comments coming from @jkbradley in #18924 , "in order to make **corpusSize**, batchSize, and nonEmptyDocsN all refer to the same filtered corpus", I think @jkbradley 's meaning is to filter out empty docs before sampling, the relationship should be:
`batchSize = miniBatchFraction * corpusSize` and `batchSize == nonEmptyDocsN`

akopich committed:
[SPARK-22111][MLLIB] OnlineLDAOptimizer should filter out empty documents beforehand

Typo fix

hhbyyh commented:
I'm curious about the performance comparison, if "filter before sample" triggers a filter over the whole dataset for each `submitMiniBatch`, then there'll be some performance impact.

And if "filter before sample" is used, IMO `miniBatchFraction`  should be adjusted proportionally.

SparkQA commented:
**[Test build #83088 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83088/testReport)** for PR 19565 at commit [`f376a7b`](https://github.com/apache/spark/commit/f376a7b97aefa700868c14da24fb6410358826df).

akopich commented:
@hhbyyh, in case of "filter before sample" in a local test the overhead is negligible. 

Regarding "sample before filter", you are right. There (strictly speaking) should be adjustment of `miniBatchFraction`. Which is why I do prefer "filter before sample".

Also note, version "sample before filter" is logically equivalent to the current upstream/master.

SparkQA commented:
**[Test build #83088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83088/testReport)** for PR 19565 at commit [`f376a7b`](https://github.com/apache/spark/commit/f376a7b97aefa700868c14da24fb6410358826df).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83088/
Test PASSed.

akopich commented:
Ping @hhbyyh, @WeichenXu123, @srowen. 

Seems like the discussion is stuck. Does anybody think that the general approach implemented in this PR should be changed? Currently it is filtering before sampling with no caching.

hhbyyh commented:
@akopich I'm actually leaning towards "filter after sample". 

1. so we don't need to change `miniBatchFraction` in 
` docs.sample(withReplacement = sampleWithReplacement, miniBatchFraction,
        randomGenerator.nextLong())`. 

2. Minor: I'm not sure corpusSize = nonEmptyDoc is always good. I'd prefer to keep docs and corpusSize referring to the entire dataset.

akopich commented:
@hhbyyh 

Yes, in this way we don't change semantics of `miniBatchFraction`. But is the way it is defined now actually correct? As I mentioned above, in the `upstram/master` the number of non-empty documents in the mini-batch is asymptotically normally distributed. Hence, the size of RDD fed to `treeAggregate` differs from one batch to another. While in this PR (filtering before sampling) all the batches have the same length.

But then again, for large datasets this should make no difference. If nobody thinks this disparity of batch sizes is an issue, I won't object against sampling before filtering.

@WeichenXu123, I believe, you were advocating for filter before sample. Do you still have the preference?

hhbyyh commented:
Let me know if I missed anything, but I don't quite catch the part 

> all the batches have the same length

 IMO
`docs.sample(withReplacement = sampleWithReplacement, miniBatchFraction, randomGenerator.nextLong())` does not return the same number of documents during multiple function calls.

akopich commented:
@hhbyyh OK, but it returns almost the same number of elements. Anyway, the variance is going to be much smaller that in the case with sample before filter.

WeichenXu123 commented:
Yes I think when dataset is large enough, using the same `miniBatchFraction`, the result RDD size of "filter before sample" and "filter after sample" will be asymptotically equal, no matter how many empty elements in dataset. (correct me if I am wrong, I am a little confused about "miniBatchFraction should be adjusted proportionally", if adjusted, then asymptotically equality is broken?)  If So, does it really effect the algo ?

akopich commented:
@WeichenXu123, in a case of large dataset this "adjustment" would have infinitesimal effect. (IMO, no adjustment is needed -- the expected number of non-empty docs in the same and does not depend on the order of filter and sample and equals to `docs.size * miniBatchFraction * fractionOfNonEmptyDocs`). 

So I believe, we all agree that sampling should go before filtering. I'll send a commit shortly.

akopich committed:
[SPARK-22111][MLLIB] OnlineLDAOptimizer should filter out empty documents beforehand

Filtering empty documents out now happens after a mini-batch is subsampled.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83301/
Test FAILed.

akopich commented:
Okay... any idea why tests failed? It says 
```ERROR: Step ?Publish JUnit test result report? failed: No test report files were found. Configuration error?```

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -497,40 +481,46 @@ final class OnlineLDAOptimizer extends LDAOptimizer with Logging {
+    val (statsSum: BDM[Double], logphatOption: Option[BDV[Double]], batchSize: Long) =
+      batch.treeAggregate((BDM.zeros[Double](k, vocabSize), logphatPartOptionBase(), 0L))({
+        case (acc, (_, termCounts)) =>
+          val stat = BDM.zeros[Double](k, vocabSize)
```
This is a per-record operation. Will it consume more memory than mapPartitions ? especially when k and vocabSize are large.

akopich replied:
Actually, we can fix this w/o falling back to `mapPartition`.

hhbyyh reviewed (commented):
Thanks for the update. About the test, probably need a manual trigger for that. cc @srowen

akopich committed:
[SPARK-22111][MLLIB] OnlineLDAOptimizer should filter out empty documents beforehand

treeAggregate is optimized. `stat` matrix is not created again for each document

akopich committed:
[SPARK-22111][MLLIB] OnlineLDAOptimizer should filter out empty documents beforehand

A new test, verifying that submitMiniBatch correctly processes an empty mini-batch, is added.

SparkQA commented:
**[Test build #83330 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83330/testReport)** for PR 19565 at commit [`8bf6f6d`](https://github.com/apache/spark/commit/8bf6f6d9d594120ed190b167c19511bcd3abf453).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83330/
Test PASSed.

akopich commented:
ping @WeichenXu123 , @srowen , @hhbyyh 
Further comments?

WeichenXu123 commented:
ok I agree this change. @jkbradley Can you take a look ?

akopich commented:
@WeichenXu123 
@jkbradley said, pings on Git don't work for him...

hhbyyh reviewed (commented):
Thanks for the update.

The code looks good to me, yet generally I would not recommend code refactor without clear benefit. Since this rewrites the main logic for `submitMiniBatch`, although probably there won't be any regression, I still suggest performance benchmark (on both speed and memory consumption with large k and vocab) on distributed cluster to make sure. 

Please change RP and jira description to reveal the code refactor. (actually maybe it's better if you can move the refactor to a separate jira).

Also I tested and it seems variationalTopicInference now can handle empty docs. But it's not verified. I created
https://issues.apache.org/jira/browse/SPARK-22502 to track it.

akopich commented:
@hhbyyh, is there a cluster I can use for this?

akopich commented:
@hhbyyh So, I guess, I should just roll the refactoring back, right?

hhbyyh commented:
It's probably better to wait for the opinion from a committer.

akopich commented:
@hhbyyh who shall we ping?

srowen reviewed (commented):
I'm OK with the change at this point if it gives a perf increase at all and doesn't change behavior. I am not sure from the discussion whether that's where we ended up.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15788/
Test PASSed.
